000X-B-C-I-R-S-error-handling-codes.md

```diff
@@ -0,0 +1,337 @@
+          "service": "translate.googleapis.com"
+        }
+      }
+    ]
+  }
+}
+```
+
+
+Below is partial table of the Standard Error code provided by gRPC and how they map to HTTP error codes. The entire list can found in the following links:
+- [gRPC Codes Table]https://cloud.google.com/apis/design/errors#handling_errors
+- (gRPC Codes ProtoBuf)[https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto]
+
+
+|HTTP | gRPC      | Description |
+|---- | ----------- | ----------- |
+|200  |	OK		  | No error. |
+|400  |	INVALID_ARGUMENT |	Client specified an invalid argument. Check error message and error details for more information. |
+|400  |	FAILED_PRECONDITION |	Request can not be executed in the current system state, such as deleting a non-empty directory. |
+|400  |	OUT_OF_RANGE |	Client specified an invalid range. |
+|401  |	UNAUTHENTICATED	| Request not authenticated due to missing, invalid, or expired OAuth token. |
+|403  |	PERMISSION_DENIED |	Client does not have sufficient permission. This can happen because the OAuth token does not have the right scopes, the client doesn't have permission, or the API has not been enabled. |
+|404  |	NOT_FOUND |	A specified resource is not found. |
+|409  |	ABORTED	| Concurrency conflict, such as read-modify-write conflict. |
+
+### Error Details Classification
+The following tables shows the propsosed error codes used in the **reason** for the **google.rpc.ErrorInfo** within the various levels of Dapr modules:
+
+**INIT**
+| Dapr Module | Description |
+| ----------- | ----------- |
+| CLI         | DPR_INIT_CLI_[1-1000] |
+| Self-hosted | DPR_INIT_SLF_[1-1000] |
+| K8S | DPR_INIT_K8S_[1-1000] |
+| Invoke | DPR_INIT_INV_[1-1000] |
+
+**RUNTIME**
+| Dapr Module | Description |
+| ----------- | ----------- |
+| CLI         | DPR_RT_CLI_[1-1000] |
+| Self-hosted | DPR_RT_SLF_[1-1000] |
+| dapr-2-dapr(HTTP) | DPR_INIT_HTTP_[1-1000] |
+| dapr-2-dapr(gRPC) | DPR_INIT_GRPC_[1-1000] |
+
+**COMPONENTS**
+| Dapr Module | Description |
+| ----------- | ----------- |
+| PubSub              | DPR_COMP_SUB_[1-1000] |
+| StateStore          | DPR_COMP_STS_[1-1000] |
+| Bindings            | DPR_COMP_BND_[1-1000] |
+| SecretStore         | DPR_COMP_SEC_[1-1000] |
+| ConfigurationStore  | DPR_COMP_CNF_[1-1000] |
+| Lock                | DPR_COMP_LCK_[1-1000] |
+| NameResolution      | DPR_COMP_NMR_[1-1000] |
+| Middleware          | DPR_COMP_MDW_[1-1000] |
+
+
+
+### Sample Error
+```json
+{
+  "code": 8089,
+  "message": "unable to perform Redis.HGETALL call: dial tcp [::1]:6379: connect: connection refused",
+  "details": [
+    {
+      "@type": "type.googleapis.com/google.rpc.ErrorInfo",
+      "reason": "DPR_COMP_STS_1",
+      "domain": "redis.state.contrib.dapr.io",
+      "metadata": {
+        "key": "myapp||name"
+      }
+    },
+    {
+      "@type": "type.googleapis.com/google.rpc.ResourceInfo",
+      "resource_type": "StateStore",
+      "resource_name": "REDIS",
```

Comment:
This is good, it means we can keep ErrorInfo agnostic of component implementation.